Bucketeer/Sources/Internal/Scheduler/TaskScheduler.swift

```diff
@@ -62,21 +89,43 @@ final class TaskScheduler {
                 object: nil
             )
         }
+
+        if .active == UIApplication.shared.applicationState {
+            // If the developer init the SDK when the app active
+            // Consider it as onForeground event()
+            onForeground()
+        }
+    }
+
+    @objc private func onAppLaunch() {
+        component.config.logger?.debug(message: "[TaskScheduler]: onAppLaunch")
+        onForeground()
+    }
+
+    @objc private func onUISceneWillConnect() {
+        component.config.logger?.debug(message: "[TaskScheduler]: onUISceneWillConnect")
+        onForeground()
     }
 
     @objc func onForeground() {
+        // noted: from iOS 16, the app will receive both UIApplication.didFinishLaunchingNotification
+        // & UIScene.willConnectNotification when the app launching
+        guard isOnForeground == false else { return }
```

Comment:
The `isOnForeground` is not thread-safe.
Also, I think we should always reschedule the foreground tasks when they change instead of ignoring them when they are already in the foreground for this case.

How about registering the notifications according to the OS version?
So we can avoid calling twice from iOS 16.
Also, I didn't find the notification `willConnectNotification` described in the comment.

Let me know your thoughts.

Reply:
I updated this logic a little, it seems we were using incorrect notifications for listening.
Currently no more duplicate events for `on_foreground` after app launched